src/msg_recorder/recorder.cc

```diff
@@ -10,25 +10,88 @@
 
 namespace cris::core {
 
-MessageRecorder::MessageRecorder(const std::filesystem::path& record_dir_prefix, std::shared_ptr<JobRunner> runner)
+MessageRecorder::MessageRecorder(const std::filesystem::path& record_dir_prefix, const int& snapshot_interval, std::shared_ptr<JobRunner> runner)
     : Base(std::move(runner))
-    , record_dir_(record_dir_prefix / RecordDirNameGenerator()) {
+    , record_dir_(record_dir_prefix / RecordDirNameGenerator())
+    , record_strand_(runner ? runner->MakeStrand() : MakeStrand())
+    , snapshot_interval_(snapshot_interval)
+    , snapshot_thread_(std::bind(&MessageRecorder::SnapshotStart, this)) {
     std::filesystem::create_directories(record_dir_);
 }
 
+void MessageRecorder::SnapshotStart() {
+    while (!snapshot_shutdown_flag_.load()) {
+        AddJobToRunner([this]() { GenerateSnapshot(keep_max_); }, record_strand_);
+        std::this_thread::sleep_for(std::chrono::seconds(snapshot_interval_));
+    }
+}
+
+void MessageRecorder::SnapshotEnd() {
+    snapshot_shutdown_flag_.store(true);
+    if (snapshot_thread_.joinable()) {
+        snapshot_thread_.join();
+    }
+    record_init_datas.clear();
+}
+
 MessageRecorder::~MessageRecorder() {
+    SnapshotEnd();
     files_.clear();
-
     if (std::filesystem::is_empty(GetRecordDir())) {
         LOG(INFO) << "Record dir " << GetRecordDir() << " is empty, removing...";
         std::filesystem::remove(GetRecordDir());
     }
 }
 
+void MessageRecorder::GenerateSnapshot(const int& max) {
+    // remove symlinks and close DB
+    for (const auto& file : files_) {
+        if (!file->GetFilePath().empty() && !std::filesystem::is_empty(file->GetFilePath())) {
+            std::filesystem::remove_all(file->GetFilePath());
```

Comment:
This removed the actual path and didn't seem like the right way to go, researching the way to remove symlink only @chenhao94

Reply:
和copy一样的原因，你遇到symlink要只处理symlink本身，默认选项可能是解引用

Reply:
不对，removeall是不删除symlink指向的内容的

> Symlinks are not followed (symlink is removed, not its target).

Reply:
Guess it's about the DB Iterator again since all errors are resolved when I only call the snapshot feature once. Maybe my current logic closed the DB before releasing the iterator pointer somehow.

Reply:
nevermind, I printed a bunch of logs and seems like it's because I didn't use the cached path to initialize the DB back, still debugging

Reply:
So in the unit test, `TestRecord()` passed but when `TestReplay(1.0)` tried to construct the `RecordFile` it failed:
```
E20221121 14:15:00.401607 506665 record_file.cc:105] RecordFile: Failed to create record file "/tmp/CRTestTmpDir.506665/record.20221121-141458.CST.pid.506665/cris__core__TestMessage_int__subid_11.ldb.d", status: NotFound: /tmp/CRTestTmpDir.506665/record.20221121-141458.CST.pid.506665/cris__core__TestMessage_int__subid_11.ldb.d/LOCK: No such file or directory

Thread 1 "msg_recorder_te" received signal SIGSEGV, Segmentation fault.
cris::core::RecordFile::Iterate (this=0x5c4d00) at src/msg_recorder/record_file.cc:139
139	    auto itr = db_->NewIterator(leveldb::ReadOptions());
(gdb) backtrace
#0  cris::core::RecordFile::Iterate (this=0x5c4d00) at src/msg_recorder/record_file.cc:139
#1  cris::core::MessageReplayer::GetRecordItr (this=0x7fffffffd2f0, message_type=..., subid=<optimized out>) at src/msg_recorder/replayer.cc:83
#2  0x00000000002228e3 in cris::core::MessageReplayer::RegisterChannel<cris::core::TestMessage<int> > (this=0x7fffffffd2f0, subid=11)
    at bazel-out/k8-opt/bin/_virtual_includes/msg_recorder/cris/core/msg_recorder/replayer.h:91
#3  0x0000000000220f6e in cris::core::RecorderTest::TestReplay (this=<optimized out>, speed_up=1) at tests/msg_recorder_test.cc:116
#4  0x00000000002235e5 in cris::core::RecorderTest_RecorderTest_Test::TestBody (this=<optimized out>) at tests/msg_recorder_test.cc:216
#5  0x000000000026f4aa in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
#6  0x000000000027d68b in testing::internal::UnitTestImpl::RunAllTests() ()
#7  0x000000000027ba44 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
#8  0x0000000000267899 in main ()
```
`files_` got cleared in the destructor, associated directories got cleared as well.